### PR TITLE
Configurable led colors issue #819 issue #2047

### PR DESCRIPTION
**Pull request by gaelj**
New CLI command: mode_color

    No arguments: lists all mode colors
    arguments: ModeIndex, ModeColorIndex, ColorIndex

First 6 groups of ModeIndexes are :

    orientation
    headfree
    horizon
    angle
    mag
    baro

Each of these groups use these ModelColorIndexes :

    north
    east
    south
    west
    up
    down

Group 6 is used for special colors according to these ModelColorIndexes:

    disarmed
    armed
    animation
    bakground

The ColorIndex is picked from the colors array ("palette").

Examples :

    set armed color to red: mode_color 6 1 2
    set disarmed color to yellow: mode_color 6 0 4
    set Headfree Mode - South to Cyan: mode_color 1 2 8

### CODE REVIEW

**Comment by gaelj** [comment]
Addresses issue #819 and issue #2047

**Comment by martinbudden** [comment]
@gaelj , as a general principle pull requests should be atomic. That is they should address one issue only. Certainly changes to GPS related things and LED related things should be in separate PRs. There are a number of reasons for this:

1. It is just good practice.
2. @hydra may choose to accept only one item, or may wish to address one or other of the items at a different time (for example, when other changes are going on in that area of code).
3. This a bigger chance of merge conflicts, which makes it much harder to accept the PR.

I would advise you to change this PR to address just the GPS issue and put the configurable colours into a separate PR.

**Comment by gaelj** [comment]
Sorry about that. I will try again :)

**Comment by hydra** [comment]
@gaelj please link this PR to your new PR or just update your PR branch with different commits and force push it over the old remote branch and then reopen this PR.  The changes are quite small and lots of people have asked for this so it's simple to merge once you give us just the commits needed.

File: src/main/io/ledstrip.c [hunk 1]
```diff
@@ -1048,6 +991,50 @@ bool parseColor(uint8_t index, const char *colorConfig)
...
 }
 
+/*
+ * Redefine a color in a mode.
+ * */
+bool setModeColor(uint8_t modeIndex, uint8_t modeColorIndex, uint8_t colorIndex)
+{
+    bool ok = true;
+
+    modeColorIndexes_t *_modeColors;
```
Comment (hydra) [inline src/main/io/ledstrip.c]: no need for `_` character on variable name, just call it what it is. :smile:

File: src/main/io/ledstrip.c [hunk 2]
```diff
@@ -1057,6 +1044,17 @@ void pgReset_colors(hsvColor_t *instance)
...
+void pgReset_modeColors(modeColorIndexes_t *instance)
+{
+    memcpy(instance, &defaultModeColors, sizeof(defaultModeColors));
+}
+
+void pgReset_specialColors(specialColorIndexes_t *instance)
+{
+    memcpy(instance, &defaultSpecialColors, sizeof(defaultSpecialColors));
+}
+
```
Comment (hydra) [inline src/main/io/ledstrip.c]: these will need updating now that the reset template code has been merged.  see other pg reset functions for details.

File: src/main/io/msp_protocol.h [hunk 3]
```diff
@@ -60,7 +60,7 @@
 #define MSP_PROTOCOL_VERSION                0
 
 #define API_VERSION_MAJOR                   1 // increment when major changes are made
-#define API_VERSION_MINOR                   17 // increment when any change is made, reset to zero when major changes are released after changing API_VERSION_MAJOR
+#define API_VERSION_MINOR                   18 // increment when any change is made, reset to zero when major changes are released after changing API_VERSION_MAJOR
```
Comment (hydra) [inline src/main/io/msp_protocol.h]: thanks for remembering this.

File: src/main/io/serial_cli.c [hunk 4]
```diff
@@ -1276,6 +1278,103 @@ static void cliColor(char *cmdline)
...
+    if (isEmpty(cmdline)) {
+        for (i = 0; i < MODE_COUNT; i++) {
+            if (i == MODE_MAG) {
+#ifdef MAG
+#else
+                    continue;
+#endif
+            }
+
+            for (j = 0; j < 6; j++) {
```
Comment (hydra) [inline src/main/io/serial_cli.c]: `6` is a magic number,  use a constant.

File: src/main/io/serial_cli.c [hunk 5]
```diff
@@ -1276,6 +1278,103 @@ static void cliColor(char *cmdline)
...
+                }
+
+                cliPrintf("mode color %u %u %u\r\n",
+                    i,
+                    j,
+                    colorIndex
+                );
+            }
+        }
+        for (j = 0; j < 4; j++) {
```
Comment (hydra) [inline src/main/io/serial_cli.c]: `4` is a magic number,  use a constant.

File: src/main/io/ledstrip.c [hunk 6]
```diff
@@ -1048,6 +991,50 @@ bool parseColor(uint8_t index, const char *colorConfig)
...
+        case MODE_HEADFREE:     _modeColors = modeColors(MODE_HEADFREE); break;
+        case MODE_HORIZON:      _modeColors = modeColors(MODE_HORIZON); break;
+        case MODE_ANGLE:        _modeColors = modeColors(MODE_ANGLE); break;
+#ifdef MAG
+        case MODE_MAG:          _modeColors = modeColors(MODE_MAG); break;
+#endif
+        case MODE_BARO:         _modeColors = modeColors(MODE_BARO); break;
+        case SPECIAL:
+            switch (modeColorIndex) {
+                case 0: specialColors(0)->disarmed = colorIndex; break;
```
Comment (hydra) [inline src/main/io/ledstrip.c]: use contants/#defines/enums for 0-3 here.

File: src/main/io/ledstrip.c [hunk 7]
```diff
@@ -1048,6 +991,50 @@ bool parseColor(uint8_t index, const char *colorConfig)
...
+                case 3: specialColors(0)->background = colorIndex; break;
+                default: return !ok;
+            }
+            return ok;
+            break;
+        default: return !ok;
+    }
+
+    switch (modeColorIndex) {
+        case 0: _modeColors->north = colorIndex; break;
```
Comment (hydra) [inline src/main/io/ledstrip.c]: use contants/#defines/enums for 0-6 here.

File: src/main/io/serial_cli.c [hunk 8]
```diff
@@ -1276,6 +1278,103 @@ static void cliColor(char *cmdline)
         }
     }
 }
+
+static void cliModeColor(char *cmdline)
+{
+    int i, j;
+    char *ptr;
+    int args[8], check = 0;
```
Comment (hydra) [inline src/main/io/serial_cli.c]: not clear why `8` is used here.

**Comment by hydra** [comment]
could you also include a commit that updates the documentation; perhaps copy/paste from PR description and update accordingly.

**Comment by gaelj** [comment]
Thanks for the comments.
New individual PR's for each of my ongoing change request coming up shortly, this time I think it got it right. Apologies for the mix-ups.

File: src/main/io/serial_cli.c (hunk 8 above)
Comment (gaelj) [inline src/main/io/serial_cli.c]: copy/paste error. You have a very good eye :)

**Comment by hydra** [comment]
@gaelj great

**Comment by gaelj** [comment]
See https://github.com/cleanflight/cleanflight/pull/2053

**Pull request closed without merging**